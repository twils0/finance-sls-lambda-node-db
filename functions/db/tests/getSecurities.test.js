const { getSecurities } = require('../getSecurities');

global.pool = {
  query: jest.fn(),
};

const userId = 'testUserID';
const securityId1 = 'testSecurityID1';
const securityId2 = 'testSecurityID2';
const securityId3 = 'testSecurityID3';
const security1 = {
  id: securityId1,
  name: 'testName1',
  exchange: 'testExchange1',
  tickerCusip: 'testTickerCusip1',
  category: 'testCategory1',
};
const security2 = {
  id: securityId2,
  name: 'testName2',
  exchange: 'testExchange2',
  tickerCusip: 'testTickerCusip2',
  category: 'testCategory2',
};
const current = securityId2;
const list = [securityId1, securityId2];
const emptySecurities = {
  current: null,
  list: [],
};
const securities = {
  current,
  list,
  [securityId1]: security1,
  [securityId2]: security2,
};
const correctedSecurities = {
  current: securityId1,
  list,
  [securityId1]: security1,
  [securityId2]: security2,
};
const emptyResponse = {
  rows: [],
};
const joinResponse = {
  rows: [
    {
      security_id: securityId1,
      security_order: 0,
    },
    {
      security_id: securityId2,
      security_order: 1,
    },
  ],
};
const securityResponse = {
  rows: [
    {
      security_id: securityId1,
      name: security1.name,
      exchange: security1.exchange,
      ticker_cusip: security1.tickerCusip,
      category: security1.category,
    },
    {
      security_id: securityId2,
      name: security2.name,
      exchange: security2.exchange,
      ticker_cusip: security2.tickerCusip,
      category: security2.category,
    },
  ],
};
const userResponse = {
  rows: [
    {
      security_current: securityId2,
    },
  ],
};
const wrongUserResponse = {
  rows: [
    {
      security_current: securityId3,
    },
  ],
};

describe('functions', () => {
  describe('db', () => {
    describe('getSecurities', () => {
      afterEach(() => {
        global.pool.query.mockReset();
      });

      it('fails and returns unexpected error, when global.pool.query throws error first query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.reject(error);
        });

        try {
          await getSecurities(userId);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
        expect(global.pool.query).toHaveBeenCalledTimes(1);
      });

      it('fails and returns unexpected error, when global.pool.query throws error second query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(joinResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([list]);
          return Promise.reject(error);
        });

        try {
          await getSecurities(userId);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
        expect(global.pool.query).toHaveBeenCalledTimes(2);
      });

      it('fails and returns unexpected error, when global.pool.query throws error second query', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(joinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([list]);
          return Promise.resolve(securityResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.reject(error);
        });

        try {
          await getSecurities(userId);
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }

        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });

      it('returns empty securities object, when security table query is empty', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(emptyResponse);
        });

        const result = await getSecurities(userId);

        expect(result).toEqual({ securities: emptySecurities });
        expect(global.pool.query).toHaveBeenCalledTimes(1);
      });

      it('returns corrected security_current, when user table query is empty', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(joinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([list]);
          return Promise.resolve(securityResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(emptyResponse);
        });

        const result = await getSecurities(userId);

        expect(result).toEqual({ securities: correctedSecurities });
        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });

      it('returns corrected security_current, when user table response has security_current not in list', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(joinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([list]);
          return Promise.resolve(securityResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(wrongUserResponse);
        });

        const result = await getSecurities(userId);

        expect(result).toEqual({ securities: correctedSecurities });
        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });

      it('returns correct securities object', async () => {
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(joinResponse);
        });
        global.pool.query.mockImplementationOnce((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([list]);
          return Promise.resolve(securityResponse);
        });
        global.pool.query.mockImplementation((dbString, valuesArray) => {
          expect(dbString).toMatchSnapshot();
          expect(valuesArray).toEqual([userId]);
          return Promise.resolve(userResponse);
        });

        const result = await getSecurities(userId);

        expect(result).toEqual({ securities });
        expect(global.pool.query).toHaveBeenCalledTimes(3);
      });
    });
  });
});
