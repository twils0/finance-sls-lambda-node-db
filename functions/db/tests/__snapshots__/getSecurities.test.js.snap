// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error first query 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error second query 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error second query 2`] = `
"SELECT security_id, name, exchange, ticker_cusip, category
        FROM security.security
        WHERE security_id IN (SELECT(UNNEST($1::uuid[])));"
`;

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error second query 3`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error second query 4`] = `
"SELECT security_id, name, exchange, ticker_cusip, category
        FROM security.security
        WHERE security_id IN (SELECT(UNNEST($1::uuid[])));"
`;

exports[`functions db getSecurities fails and returns unexpected error, when global.pool.query throws error second query 5`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns correct securities object 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns correct securities object 2`] = `
"SELECT security_id, name, exchange, ticker_cusip, category
        FROM security.security
        WHERE security_id IN (SELECT(UNNEST($1::uuid[])));"
`;

exports[`functions db getSecurities returns correct securities object 3`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns corrected security_current, when user table query is empty 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns corrected security_current, when user table query is empty 2`] = `
"SELECT security_id, name, exchange, ticker_cusip, category
        FROM security.security
        WHERE security_id IN (SELECT(UNNEST($1::uuid[])));"
`;

exports[`functions db getSecurities returns corrected security_current, when user table query is empty 3`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns corrected security_current, when user table response has security_current not in list 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns corrected security_current, when user table response has security_current not in list 2`] = `
"SELECT security_id, name, exchange, ticker_cusip, category
        FROM security.security
        WHERE security_id IN (SELECT(UNNEST($1::uuid[])));"
`;

exports[`functions db getSecurities returns corrected security_current, when user table response has security_current not in list 3`] = `
"SELECT security_current FROM \\"user\\".\\"user\\"
        WHERE user_id = $1::uuid;"
`;

exports[`functions db getSecurities returns empty securities object, when security table query is empty 1`] = `
"SELECT security_id, security_order
        FROM \\"user\\".\\"join_user_security\\"
        WHERE user_id = $1::uuid;"
`;
